File: samples/OvalSample/OvalSample/ViewController.m
//
//  ViewController.m
//  OvalSample
//
//  Created by Corey Earwood on 6/23/15.
//  Copyright (c) 2015 Orbotix. All rights reserved.
//

#import "ViewController.h"
#import <RobotKit/RobotKit.h>

@interface ViewController () <RKOvalControlDelegate>

@property (nonatomic, weak) IBOutlet UITextField *lightSpeed;
@property (nonatomic, strong) RKConvenienceRobot *robot;
@property (nonatomic, strong) RKOvalControl *ovalControl;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appDidBecomeActive:) name:UIApplicationDidBecomeActiveNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appWillResignActive:) name:UIApplicationWillResignActiveNotification object:nil];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)updateOval:(id)sender {
    [_ovalControl sendOvalString:[NSString stringWithFormat:@"speed = %@;...",_lightSpeed.text]];
}

- (void)appDidBecomeActive:(NSNotification *)note {
    [[RKRobotDiscoveryAgent sharedAgent] addNotificationObserver:self selector:@selector(handleRobotStateChangeNotification:)];
    [RKRobotDiscoveryAgent startDiscovery];
}

- (void)appWillResignActive:(NSNotification *)note {
    [RKRobotDiscoveryAgent disconnectAll];
}

- (void)handleRobotStateChangeNotification:(RKRobotChangedStateNotification *)notification {
    switch (notification.type) {
        case RKRobotOnline:
            self.robot = [[RKConvenienceRobot alloc] initWithRobot:notification.robot];
            self.ovalControl = [[RKOvalControl alloc] initWithRobot:notification.robot delegate:self];
            [self.ovalControl resetOvmAndLibrary:YES];
            [self sendOvalProgram];
            break;
            
        default:
            break;
    }
}

- (void)sendOvalProgram {
    NSString *ovalProgram = [NSString stringWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"Sample" ofType:@"oval"] encoding:NSUTF8StringEncoding error:nil];
    [_ovalControl sendOvalPrograms:@[ovalProgram]];
}

#pragma mark - RKOvalControlDelegate

- (void)ovalControlDidFinishSendingProgram:(RKOvalControl *)control {
    NSLog(@"Oval successfully sent");
}

- (void)ovalControlDidResetOvm:(RKOvalControl *)control {
    NSLog(@"OVM Reset");
}

- (void)ovalControl:(RKOvalControl *)control receivedOvalNotification:(RKOvalDeviceBroadcast *)notification {
    NSLog(@"Did receive oval async with floats: %@ ints: %@",notification.floats, notification.ints);
}

- (void)ovalControl:(RKOvalControl *)control receivedVmRuntimeError:(RKOvalErrorBroadcast *)notification {
    NSLog(@"Did receive OVM Error: %@",notification.errorDescription);
}

- (void)ovalControl:(RKOvalControl *)control didFailToSendProgramWithMessage:(NSString *)message {
    NSLog(@"Failed tos end program with message: %@",message);
}

@end

File: samples/OvalSampleSwift/OvalSampleSwift/ViewController.swift
//
//  ViewController.swift
//  OvalSampleSwift
//
//  Created by Corey Earwood on 6/22/15.
//  Copyright (c) 2015 Orbotix. All rights reserved.
//

import UIKit

class ViewController: UIViewController, RKOvalControlDelegate, UITextFieldDelegate {
    var robot: RKConvenienceRobot!
    var ovalControl: RKOvalControl!
    @IBOutlet var lightSpeed: UITextField!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let tap = UITapGestureRecognizer(target:self, action:#selector(hideKeyboard));
        tap.cancelsTouchesInView = false;
        self.view.addGestureRecognizer(tap);
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        RKRobotDiscoveryAgent.shared().addNotificationObserver(self, selector: #selector(handleRobotStateChangeNotification(_:)))
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        NotificationCenter.default.addObserver(self, selector: #selector(appDidBecomeActive(_:)), name: NSNotification.Name.UIApplicationDidBecomeActive, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(appWillResignActive(_:)), name: NSNotification.Name.UIApplicationWillResignActive, object: nil)
    }
    
    func appDidBecomeActive(_: Notification) {
        RKRobotDiscoveryAgent.startDiscovery()
    }
    
    func appWillResignActive(_: Notification) {
        RKRobotDiscoveryAgent.disconnectAll()
    }

    func handleRobotStateChangeNotification(_ notification: RKRobotChangedStateNotification) {
        switch (notification.type) {
        case .online:
            self.robot = RKConvenienceRobot(robot: notification.robot)
            self.ovalControl = RKOvalControl(robot: notification.robot, delegate: self)
        case .disconnected:
            self.robot = nil
            self.ovalControl = nil
        default:
            break
        }
    }
    
    func sendOvalProgram() {
        guard let path = Bundle.main.path(forResource: "Sample", ofType: "oval") else {
            return
        }
        
        var source: String? = nil
        do {
            source = try String(contentsOfFile: path, encoding: String.Encoding.utf8)
            if let unwrappedSource = source {
                self.ovalControl.sendOvalPrograms([unwrappedSource])
            }
        } catch let error as NSError {
            debugPrint("Error: \(error)")
        }
    }
    
    func hideKeyboard() {
        self.view.endEditing(true);
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        hideKeyboard();
        return true;
    }
    
    @IBAction func updateOval(_ sender: AnyObject?) {
        self.ovalControl.sendOvalString("speed = \(lightSpeed.text!);...")
    }
    
    //MARK: RKOvalControlDelegate
    
    func ovalControlDidInitialize(_ control: RKOvalControl!) {
        self.ovalControl.resetOvmAndLibrary(true)
    }
    
    func ovalControlDidFinishSendingProgram(_ control: RKOvalControl!) {
        debugPrint("Oval successfully sent")
    }
    
    func ovalControlDidResetOvm(_ control: RKOvalControl!) {
        debugPrint("OVM Reset")
        sendOvalProgram()
    }
    
    func ovalControl(_ control: RKOvalControl!, receivedOvalNotification notification: RKOvalDeviceBroadcast!) {
        debugPrint("Did receive oval async with floats: \(notification.floats) ints: \(notification.ints)")
    }
    
    func ovalControl(_ control: RKOvalControl!, receivedVmRuntimeError notification: RKOvalErrorBroadcast!) {
        debugPrint("Did receive OVM Error: \(notification.errorDescription())")
    }
    
    func ovalControl(_ control: RKOvalControl!, didFailToSendProgramWithMessage message: String!) {
        debugPrint("Failed to send program with message: \(message)")
    }
}

